### DISCUSSION TITLE
Help in creating hybrid ranks

### DISCUSSION THREAD

**Discussion by davidgritter** [GrandOrgue/GoOdf#51] [General]
I am trying to make a hybrid rank, that is some of the pipes come from a different directory than the others.  The first twelve pipes come from  one directory who's first pipe is midi note number 24.  So if I set the number of pipes to 61, and the first midi note number to 24, I can load the 12 samples from that directory into  pipes 1 thru 12.  For the second directory I can change the first midi note number to 12, and these samples will load correctly into pipes 13-61 if I click yes in the popup after changed first midi note number.  Unfortunately, clicking yes also moves the samples from the first 12 pipes to start at pipe number 13.  When I click no in the popup, the new samples load starting at pipe1, and lie on top of the first 12 samples.  This appears to be the case both when I use create new pipes from midi nbr, or use add more samples from..   Can youi tell me if there is a proper way to displace added samples to higher pipe numbers without also displacing samples already loaded into lower pipe numbers?

**Comment by Bumblebee001** [comment]
Whilst the question is very interesting in as of itself, I would have
simply copied all the wav files I need into a common folder and referenced
the path of that folder.

**Comment by larspalo** [comment]
Well, for ranks the First MIDI note number is essential and has a real use - so it must be correct for the situation. For the internal rank in a stop it's not really critical as that value is derived from the manual (owning the stop). Nevertheless, I'm a bit concerned about it. I really think that the value should be set to what's appropriate for the rank/stop and be left as it is (not changed).

For more flexibility in the (auto) loading of pipes I'm considering a new dialog option where the loading parameters can be variable without affecting, or using, the global parameters at all. Instead one could specify in that dialog from what pipe to start and what MIDI note number to look for for that pipe, then exactly how many pipes to load and with what options... This will take time to implement though.

I'm also considering a new possibility to copy the attack(s)/release(s), and auto adjust PitchTuning properties automatically from the source used in the copy, to whatever pipes is selected - not necessarily in sequential order. This would facilitate extension of ranks/stops. Presently, the copy options are somewhat limited to only the pipes coming next after the source.

But, for the time being, your best bet is to auto load pipes for the most of them, and then manually add the required extra pipes. Or do as Mark suggested and make a new sample set folder where you copy all used material at the expense of higher disk usage...

**Comment by davidgritter** [comment]
I am not really happy with the idea of moving samples to different directories:

- The resulting .organ file becomes not portable to someone else who may have the same commercial sample set unless they too move samples
- the samples not only need to be moved but renamed.  In my case, for example I would have two samples each for midi notes 24 to 35, which GOODF would load as two attacks each for pipes 1 to 11  renaming would need to be done for all attacks and all releases for both normal and tremmed samples

I note that jmakeodf had a very nice mechanism for solving this problem, which involved a separate line for different length  groups of pipes, where the sum total of samples loaded had to equal the total number of pipes.  This seems to be similar to what Lars is proposing.

I also note that  a simple change to the rank dialogs in GOODF could accomplish the same thing:   If one loads a set of samples, the first midi note number is used to align samples to pipes.   If, after loading some samples, it is changed, the yes/no dialog box pops up.  If one answers yes, existing samples are moved to different pipes.  If, one answers no, it seems the newly entered first midi note number has no effect on the loading of additional samples, even though it remains as newly entered on the rank element panel.  If instead, when one answered no, the existing samples were not moved, but subsequently loaded samples used the new value of midi note number to determine which pipe they should be attached to, the problem would be solved.

**Comment by davidgritter** [comment]
For now what I've done is created two ranks, one with the first twelve pipes and the second with the top 48. they are both referenced by the same stop

**Reply by larspalo** [reply]
> For now what I've done is created two ranks

@davidgritter Yes, that's a nice solution too, and actually a good use case for referencing ranks over simple stop(s)!

For the original problem, I'm very much leaning towards a new pipe loading/reading method that would be more flexible and possible to configure without affecting any other settings.

**Reply by larspalo** [reply]
With the 0.8.0 release there's now another button for more flexible automatic pipe loading like outlined above.
